File: appcenter/ios/Plugin/AppCenterBase.swift
import Foundation
import Capacitor
import AppCenter

@objc public class AppCenterBase: NSObject {

    public func isNetWorkRequestsAllowed() -> Bool {
        return AppCenter.networkRequestsAllowed
    }

    public func setNetworkRequestsAllowed(_ shouldAllow: Bool) {
        AppCenter.networkRequestsAllowed = shouldAllow
    }

    public func setLogLevel(_ level: Int) {
        AppCenter.logLevel = LogLevel.init(rawValue: UInt(level)) ?? .verbose
    }

    public func getLogLevel() -> UInt {
        return AppCenter.logLevel.rawValue
    }

    public func getInstallId() -> String {
        return AppCenter.installId.uuidString
    }

    public func getSdkVersion() -> String {
        return AppCenter.sdkVersion
    }

    public func setUserId(_ id: String) {
        AppCenter.userId = id
    }

    public func enable(_ flag: Bool) {
        AppCenter.enabled = flag
    }

    public func isEnabled() -> Bool {
        return AppCenter.enabled
    }

    public func setCustomProperties(_ properties: [String: [String: Any]]) {
        let customProperties = CustomProperties()

        for (propName, propData) in properties {

            if let type = propData["type"] as? String {
                
                if type == "string", let value = propData["value"] as? String {
                    customProperties.set(value, forKey: propName)
                } else if type == "number", let value = propData["value"] as? NSNumber {
                    customProperties.set(value, forKey: propName)
                } else if type == "boolean", let value = propData["value"] as? Bool {
                    customProperties.set(value, forKey: propName)
                } else if type == "date-time", let value = propData["value"] as? Double {
                    let date = Date.init(timeIntervalSince1970: value)
                    customProperties.set(date, forKey: propName)
                } else if type == "clear" {
                    customProperties.clearProperty(forKey: propName)
                }
            }
        }

        AppCenter.setCustomProperties(customProperties)
    }

}

File: appcenter/android/src/main/java/com/getcapacitor/plugin/appcenter/AppCenterBase.java
package com.getcapacitor.plugin.appcenter;

import com.getcapacitor.JSObject;
import com.microsoft.appcenter.AppCenter;
import com.microsoft.appcenter.CustomProperties;
import java.util.Date;
import java.util.Iterator;
import java.util.UUID;
import org.json.JSONObject;

public class AppCenterBase {

    public void setNetworkRequestsAllowed(boolean shouldAllow) {
        AppCenter.setNetworkRequestsAllowed(shouldAllow);
    }

    public boolean isNetworkRequestsAllowed() {
        return AppCenter.isNetworkRequestsAllowed();
    }

    public void setLogLevel(int level) {
        AppCenter.setLogLevel(level);
    }

    public int getLogLevel() {
        return AppCenter.getLogLevel();
    }

    public String getInstallId() {
        UUID installId = AppCenter.getInstallId().get();
        return installId == null ? null : installId.toString();
    }

    public String getSdkVersion() {
        return AppCenter.getSdkVersion();
    }

    public void setUserId(String userId) {
        AppCenter.setUserId(userId);
    }

    public void enable(boolean enabled) {
        AppCenter.setEnabled(enabled).get();
    }

    public boolean isEnabled() {
        return AppCenter.isEnabled().get();
    }

    public void setCustomProperties(JSObject properties) {
        CustomProperties customProps = new CustomProperties();
        Iterator<String> keysIter = properties.keys();

        while (keysIter.hasNext()) {
            String key = keysIter.next();
            // JSONObject valueObject = (JSONObject) properties.getJSObject(key);
            JSObject valueObject = properties.getJSObject(key);

            if (valueObject != null) {
                String type = valueObject.getString("type");
                switch (type) {
                    case "clear":
                        customProps.clear(key);
                        break;
                    case "string":
                        customProps.set(key, valueObject.getString("value"));
                        break;
                    case "number":
                        customProps.set(key, Double.parseDouble(valueObject.getString("value")));
                        break;
                    case "boolean":
                        customProps.set(key, valueObject.getBool("value"));
                        break;
                    case "date-time":
                        customProps.set(key, new Date((long) Double.parseDouble(valueObject.getString("value"))));
                        break;
                }
            }
        }

        AppCenter.setCustomProperties(customProps);
    }
}
